libs/remote_storage/src/s3_bucket.rs

```diff
@@ -569,64 +624,138 @@ impl RemoteStorage for S3Bucket {
-                        // would be too verbose, especially as errors may lead us
-                        // to retry repeatedly.
-                        const LOG_UP_TO_N_ERRORS: usize = 10;
-                        for e in errors.iter().take(LOG_UP_TO_N_ERRORS) {
-                            tracing::warn!(
-                                "DeleteObjects key {} failed: {}: {}",
-                                e.key.as_ref().map(Cow::from).unwrap_or("".into()),
-                                e.code.as_ref().map(Cow::from).unwrap_or("".into()),
-                                e.message.as_ref().map(Cow::from).unwrap_or("".into())
-                            );
-                        }
-
-                        return Err(anyhow::format_err!(
-                            "Failed to delete {} objects",
-                            errors.len()
-                        ));
+            hm_vd
+                .entry(key)
+                .or_default()
+                .push((vd, last_modified, version_id));
+        }
+        for (key, versions) in hm_vd {
+            let (last_vd, last_last_modified, _version_id) = versions.last().unwrap();
+            if last_last_modified > &&done_if_after {
+                tracing::info!("Key {key} has version later than done_if_after, skipping");
+                continue;
+            }
+            // the version we want to restore to.
+            let version_to_restore_to =
+                match versions.binary_search_by_key(&timestamp, |tpl| *tpl.1) {
+                    Ok(v) => v,
+                    Err(e) => e,
+                };
+            if version_to_restore_to == versions.len() {
+                tracing::info!("Key {key} has no changes since timestamp, skipping");
+                continue;
+            }
+            let mut do_delete = false;
+            if version_to_restore_to == 0 {
+                // All versions more recent, so the key didn't exist at the specified time point.
+                tracing::info!(
+                    "All {} versions more recent for {key}, deleting",
+                    versions.len()
+                );
+                do_delete = true;
+            } else {
+                match &versions[version_to_restore_to - 1] {
+                    (VerOrDelete::Version(_), _last_modified, version_id) => {
+                        tracing::info!("Copying old version {version_id} for {key}...");
+                        // Restore the state to the last version by copying
+                        let source_id =
+                            format!("{}/{key}?versionId={version_id}", self.bucket_name);
+
+                        self.client
+                            .copy_object()
+                            .bucket(self.bucket_name.clone())
+                            .key(key)
+                            .copy_source(source_id)
+                            .send()
+                            .await?;
```

Comment:
Shouldn't we do `index_part.json` files after we've restored the rest of the state at the desired timestamp? My concern is ending up with an index that points to layers which do not exist if this operation fails midway through.

Reply:
This is the low level side of things in the remote_storage module. It knows nothing about the pageserver or other components. 

The way I'd handle a case where the operation fails midway is to just retry it. The operation is designed in a way that it remembers progress, that's what the `limit_timestamp` parameter is for.

Also, on the higher level, one can use this low level operation to specify an arbitrary prefix, including the path of index_part.json. In other words, if one wants, one can restore index_part.json to its old state before doing any other restoration operations.